cosmos/x/evm/plugins/txpool/mempool/mempool.go

```diff
@@ -22,89 +22,192 @@ package mempool
 
 import (
 	"context"
+	"sync"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
-	sdkmempool "github.com/cosmos/cosmos-sdk/types/mempool"
 
+	cosmlib "pkg.berachain.dev/polaris/cosmos/lib"
 	"pkg.berachain.dev/polaris/cosmos/x/evm/types"
 	"pkg.berachain.dev/polaris/eth/common"
 	coretypes "pkg.berachain.dev/polaris/eth/core/types"
 	"pkg.berachain.dev/polaris/lib/utils"
 )
 
-// EthTxPool is a mempool for Ethereum transactions. It wraps a
-// PriorityNonceMempool and caches transactions that are added to the mempool by
-// ethereum transaction hash.
+// EthTxPool is a mempool for Ethereum transactions. It wraps a PriorityNonceMempool and caches
+// transactions that are added to the mempool by ethereum transaction hash.
 type EthTxPool struct {
-	sdkmempool.Mempool
+	// The underlying mempool implementation.
+	*PriorityNonceMempool[int64]
 
-	// ethTxCache caches transactions that are added to the mempool
-	// so that they can be retrieved later
+	// ethTxCache caches transactions that are added to the mempool so that they can be retrieved
+	// later
 	ethTxCache map[common.Hash]*coretypes.Transaction
 
-	// // nonceCache caches the pending nonce by txhash
-	// nonceCache map[common.Address]*coretypes.Transaction
-
-	// // minedBlockCache caches the mined transaction by block hash
-	// minedBlockCache map[common.Hash][]*coretypes.Transaction
-
-	// blockNumberCache
+	// NonceRetriever is used to retrieve the nonce for a given address (this is typically a
+	// reference to the StateDB).
+	nr NonceRetriever
 
+	// We have a mutex to protect the ethTxCache and nonces maps since they are accessed
+	// concurrently by multiple goroutines.
+	mu sync.RWMutex
 }
 
 // New is called when the mempool is created.
-func NewEthTxPoolFrom(m sdkmempool.Mempool) *EthTxPool {
+func NewEthTxPoolFrom(mp *PriorityNonceMempool[int64]) *EthTxPool {
 	return &EthTxPool{
-		Mempool:    m,
-		ethTxCache: make(map[common.Hash]*coretypes.Transaction),
+		PriorityNonceMempool: mp,
+		ethTxCache:           make(map[common.Hash]*coretypes.Transaction),
 	}
 }
 
+// SetNonceRetriever sets the nonce retriever db for the mempool.
+func (etp *EthTxPool) SetNonceRetriever(nr NonceRetriever) {
+	etp.nr = nr
+}
+
 // Insert is called when a transaction is added to the mempool.
 func (etp *EthTxPool) Insert(ctx context.Context, tx sdk.Tx) error {
+	etp.mu.Lock()
+	defer etp.mu.Unlock()
+
 	// Call the base mempool's Insert method
-	if err := etp.Mempool.Insert(ctx, tx); err != nil {
+	if err := etp.PriorityNonceMempool.Insert(ctx, tx); err != nil {
 		return err
 	}
 
 	// We want to cache
-	etr, ok := utils.GetAs[*types.EthTransactionRequest](tx.GetMsgs()[0])
-	if !ok {
-		return nil
+	if ethTx := GetAsEthTx(tx); ethTx != nil {
```

Comment:
GetAsEthTx maybe should go in evm/types?

Reply:
Actually you can cast the sdk.Tx to EthTransactionRequest and then use .AsTransaction()